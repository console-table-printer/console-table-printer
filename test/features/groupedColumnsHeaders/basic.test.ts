import { Table } from '../../../index';
import { GroupedColumnsHeaderOrPlaceholder } from '../../../src/models/external-table';
import { getTableBody, getTableHeader } from '../../testUtils/getRawData';

describe('Grouped columns headers', () => {
  const buildTestTable = (
    groupedHeaders?: GroupedColumnsHeaderOrPlaceholder[]
  ): Table => {
    const table = new Table({
      groupedColumnsHeaders: groupedHeaders,
    });

    table.addRow({
      A: `A1`,
      B: `B1`,
      C: `C1`,
      D: `D1`,
      E: `E1`,
      F: `F1`,
      G: `G1`,
      H: `H1`,
      I: `I 00001`,
    });

    table.addRow({
      A: `A2`,
      B: `B2`,
      C: `C2`,
      D: `D 00002`,
      E: `E2`,
      F: `F2`,
      G: `G2`,
      H: `H2`,
      I: `I2`,
    });

    return table;
  };

  it('should render single group', () => {
    const p = buildTestTable([{ name: 'G1', width: 9 }]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should render consecutive groups', () => {
    const p = buildTestTable([
      { name: 'G1', width: 2 },
      { name: 'G2', width: 2 },
      { name: 'G3', width: 5 },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should render consecutive groups with a placeholder at the start', () => {
    const p = buildTestTable([
      { kind: 'PLACEHOLDER', width: 1 },
      { name: 'G1', width: 2 },
      { name: 'G2', width: 2 },
      { name: 'G3', width: 4 },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should render consecutive groups with a placeholder at the end', () => {
    const p = buildTestTable([
      { name: 'G1', width: 2 },
      { name: 'G2', width: 2 },
      { name: 'G3', width: 4 },
      { kind: 'PLACEHOLDER', width: 1 },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should render consecutive groups with a placeholder in between', () => {
    const p = buildTestTable([
      { name: 'G1', width: 2 },
      { name: 'G2', width: 2 },
      { kind: 'PLACEHOLDER', width: 1 },
      { name: 'G3', width: 4 },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should render consecutive groups with placeholders', () => {
    const p = buildTestTable([
      { kind: 'PLACEHOLDER', width: 1 },
      { name: 'G1', width: 2 },
      { kind: 'PLACEHOLDER', width: 1 },
      { name: 'G2', width: 2 },
      { name: 'G3', width: 2 },
    ]);

    p.printTable();

    // ensure we didn't mutate the original input when normalizing
    expect(p.table.groupedColumnsHeaders.length).toBe(5);

    expect(p.render()).toMatchSnapshot();
  });

  it('should render placeholders spanning multiple columns', () => {
    const p = buildTestTable([
      { kind: 'PLACEHOLDER', width: 2 },
      { kind: 'PLACEHOLDER', width: 3 },
      { name: 'G1', width: 2 },
    ]);

    p.printTable();

    // ensure we don't mutate the original input when normalizing
    expect(p.table.groupedColumnsHeaders.length).toBe(3);

    expect(p.render()).toMatchSnapshot();
  });

  it('should cope with very long group names', () => {
    // Note: Current rendering strategy produces the following output
    // where the group name is truncated to fit the width of spanned columns:
    //
    //                                         ┌──────────────┐
    //                                         │  VERY LONG   │
    // ┌────┬────┬────┬─────────┬────┬────┬────┼────┬─────────┤
    // │  A │  B │  C │       D │  E │  F │  G │  H │       I │
    // ├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
    // │ A1 │ B1 │ C1 │      D1 │ E1 │ F1 │ G1 │ H1 │ I 00001 │
    // │ A2 │ B2 │ C2 │ D 00002 │ E2 │ F2 │ G2 │ H2 │      I2 │
    // └────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘
    //
    // An alternative rendering strategy could be put in place to produce the following:
    // However, it would be far more involved to implement.
    //
    //                                         ┌──────────────────────┐
    //                                         │ VERY LONG GROUP NAME │
    // ┌────┬────┬────┬─────────┬────┬────┬────┼─────────┬────────────┤
    // │  A │  B │  C │       D │  E │  F │  G │       H │          I │
    // ├────┼────┼────┼─────────┼────┼────┼────┼─────────┼────────────┤
    // │ A1 │ B1 │ C1 │      D1 │ E1 │ F1 │ G1 │      H1 │    I 00001 │
    // │ A2 │ B2 │ C2 │ D 00002 │ E2 │ F2 │ G2 │      H2 │         I2 │
    // └────┴────┴────┴─────────┴────┴────┴────┴─────────┴────────────┘

    const p = buildTestTable([
      { kind: 'PLACEHOLDER', width: 7 },
      { name: 'VERY LONG GROUP NAME', width: 2 },
    ]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should ignore headers only made of placeholders', () => {
    const p = buildTestTable([
      { kind: 'PLACEHOLDER', width: 2 },
      { kind: 'PLACEHOLDER', width: 3 },
    ]);

    p.printTable();

    // ensure we didn't mutate the original input when normalizing
    expect(p.table.groupedColumnsHeaders.length).toBe(2);

    expect(p.render()).toMatchSnapshot();
  });

  describe('should throw when total table width is narrower than total width of group headers', () => {
    it('ending with a group header', () => {
      const p = buildTestTable([
        { kind: 'PLACEHOLDER', width: 3 },
        { name: 'G1', width: 2 },
        { kind: 'PLACEHOLDER', width: 1 },
        { name: 'G2', width: 2 },
        { name: 'G3', width: 2 },
      ]);

      expect(() => p.render()).toThrowErrorMatchingSnapshot();
    });

    it('ending with a placeholder', () => {
      const p = buildTestTable([
        { kind: 'PLACEHOLDER', width: 3 },
        { name: 'G1', width: 2 },
        { kind: 'PLACEHOLDER', width: 1 },
        { name: 'G2', width: 2 },
        { kind: 'PLACEHOLDER', width: 2 },
      ]);

      expect(() => p.render()).toThrowErrorMatchingSnapshot();
    });

    it('being one too large group header', () => {
      const p = buildTestTable([{ name: 'G1', width: 99 }]);

      expect(() => p.render()).toThrowErrorMatchingSnapshot();
    });

    it('being one too large placeholder', () => {
      const p = buildTestTable([{ kind: 'PLACEHOLDER', width: 99 }]);

      expect(() => p.render()).toThrowErrorMatchingSnapshot();
    });
  });
});
