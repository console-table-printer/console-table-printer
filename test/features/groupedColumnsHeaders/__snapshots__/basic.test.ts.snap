// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Grouped columns headers should cope with very long group names 1`] = `
"                                        ┌──────────────┐
 [37m [0m[01m                                     [0m │[37m [0m[01m VERY LONG  [0m │
┌────┬────┬────┬─────────┬────┬────┬────┼────┬─────────┤
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should ignore headers only made of placeholders 1`] = `
"┌────┬────┬────┬─────────┬────┬────┬────┬────┬─────────┐
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should render consecutive groups 1`] = `
"┌─────────┬──────────────┬─────────────────────────────┐
│[37m [0m[01m  G1   [0m │[37m [0m[01m     G2     [0m │[37m [0m[01m            G3             [0m │
├────┬────┼────┬─────────┼────┬────┬────┬────┬─────────┤
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should render consecutive groups with a placeholder at the end 1`] = `
"┌─────────┬──────────────┬───────────────────┐
│[37m [0m[01m  G1   [0m │[37m [0m[01m     G2     [0m │[37m [0m[01m       G3        [0m │[37m [0m[01m       [0m  
├────┬────┼────┬─────────┼────┬────┬────┬────┼─────────┐
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should render consecutive groups with a placeholder at the start 1`] = `
"     ┌─────────┬──────────────┬────────────────────────┐
 [37m [0m[01m  [0m │[37m [0m[01m  G1   [0m │[37m [0m[01m     G2     [0m │[37m [0m[01m          G3          [0m │
┌────┼────┬────┼─────────┬────┼────┬────┬────┬─────────┤
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should render consecutive groups with a placeholder in between 1`] = `
"┌─────────┬──────────────┐    ┌────────────────────────┐
│[37m [0m[01m  G1   [0m │[37m [0m[01m     G2     [0m │[37m [0m[01m  [0m │[37m [0m[01m          G3          [0m │
├────┬────┼────┬─────────┼────┼────┬────┬────┬─────────┤
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should render consecutive groups with placeholders 1`] = `
"     ┌─────────┐         ┌─────────┬─────────┐
 [37m [0m[01m  [0m │[37m [0m[01m  G1   [0m │[37m [0m[01m       [0m │[37m [0m[01m  G2   [0m │[37m [0m[01m  G3   [0m │[37m [0m[01m       [0m  
┌────┼────┬────┼─────────┼────┬────┼────┬────┼─────────┐
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should render placeholders spanning multiple columns 1`] = `
"                              ┌─────────┐
 [37m [0m[01m                           [0m │[37m [0m[01m  G1   [0m │[37m [0m[01m            [0m  
┌────┬────┬────┬─────────┬────┼────┬────┼────┬─────────┐
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should render single group 1`] = `
"┌──────────────────────────────────────────────────────┐
│[37m [0m[01m                         G1                         [0m │
├────┬────┬────┬─────────┬────┬────┬────┬────┬─────────┤
│[37m [0m[01m A[0m │[37m [0m[01m B[0m │[37m [0m[01m C[0m │[37m [0m[01m      D[0m │[37m [0m[01m E[0m │[37m [0m[01m F[0m │[37m [0m[01m G[0m │[37m [0m[01m H[0m │[37m [0m[01m      I[0m │
├────┼────┼────┼─────────┼────┼────┼────┼────┼─────────┤
│[37m [0m[37mA1[0m │[37m [0m[37mB1[0m │[37m [0m[37mC1[0m │[37m [0m[37m     D1[0m │[37m [0m[37mE1[0m │[37m [0m[37mF1[0m │[37m [0m[37mG1[0m │[37m [0m[37mH1[0m │[37m [0m[37mI 00001[0m │
│[37m [0m[37mA2[0m │[37m [0m[37mB2[0m │[37m [0m[37mC2[0m │[37m [0m[37mD 00002[0m │[37m [0m[37mE2[0m │[37m [0m[37mF2[0m │[37m [0m[37mG2[0m │[37m [0m[37mH2[0m │[37m [0m[37m     I2[0m │
└────┴────┴────┴─────────┴────┴────┴────┴────┴─────────┘"
`;

exports[`Grouped columns headers should throw when total table width is narrower than total width of group headers being one too large group header 1`] = `"Grouped columns width (99) exceeds total columns width (9)"`;

exports[`Grouped columns headers should throw when total table width is narrower than total width of group headers being one too large placeholder 1`] = `"Grouped columns width (99) exceeds total columns width (9)"`;

exports[`Grouped columns headers should throw when total table width is narrower than total width of group headers ending with a group header 1`] = `"Grouped columns width (10) exceeds total columns width (9)"`;

exports[`Grouped columns headers should throw when total table width is narrower than total width of group headers ending with a placeholder 1`] = `"Grouped columns width (10) exceeds total columns width (9)"`;
